platform/logger.h

```diff
@@ -0,0 +1,150 @@
+#define GET_MODULE_ID(x)        (GET_MODULE_ID_(x[0], x[1], x[2], x[3]))
+
+#define SET_MODULE(x)           (GET_MODULE_ID(x) & 0xFF)
+#define SET_COUNTER(y)          ((y & 0xFF) << 8)
+#define SET_LINE_NUM(z)         ((z & 0xFFFF) << 16)
+#define TRACE_ID_(x,y,z)        (SET_MODULE(x) | SET_COUNTER(y) | SET_LINE_NUM(z))
+
+// Macros to log ID based data
+#define MBED_LOG_ID_4(...)                            log_buffer_id_data(__VA_ARGS__)
+// Data dumped in special section is : Unique ID, format string length and format string
+#if 0
+#define MBED_LOG_ID_3(counter, id, args, fmt, ...)    ({volatile static const __attribute__((section(".keep.log_data"))) char str##counter[] = fmt; \
+                                                       volatile static const __attribute__((section(".keep.log_data"))) uint32_t len##counter = MBED_STRLEN(fmt); \
+                                                       volatile static const __attribute__((section(".keep.log_data"))) uint32_t c##counter = id; \
+                                                       MBED_LOG_ID_4(args, id, ##__VA_ARGS__); \
+                                                      })
+#else
+#define MBED_LOG_ID_3(counter, id, args, fmt, ...)    MBED_LOG_ID_4(args, id, ##__VA_ARGS__);
+#endif
+
+#define MBED_LOG_ID_2(counter, id, ...)               MBED_LOG_ID_3(counter, id, MBED_COUNT_VA_ARGS(__VA_ARGS__), ##__VA_ARGS__)
+#define MBED_LOG_ID_1(mod, fmt, ll, f, l, c, ...)     MBED_LOG_ID_2(c, TRACE_ID_(mod,c,l), "[" mod "][" f "][" MBED_STRINGIFY(l) "]: " fmt, ##__VA_ARGS__)
+
+// Macros to log string data
+#define MBED_LOG_STR(...)                             log_buffer_string_data(__VA_ARGS__)
+#define MBED_LOG_STR_1(mod, fmt, ll, ...)             MBED_LOG_STR("[%-4.4s][%-4.4s]: " fmt, ll, mod, ##__VA_ARGS__)
+
+// Assert string
+#define MBED_LOG_ASSERT(...)                          log_assert(__VA_ARGS__)
+#define MBED_LOG_ASSERT_1(mod, fmt, ll, f, l, ...)    MBED_LOG_ASSERT("[%-4.4s][%-4.4s][%-15s][%5d]: " fmt, ll, mod, f, l, ##__VA_ARGS__)
+
+#define MBED_LOG_VSTR_2(fmt1, ap)                     log_buffer_string_vdata(fmt1, ap)
+#define MBED_LOG_VSTR_1(ll, mod, fmt, ap)             MBED_LOG_VSTR_2("[" ll "][" mod "]: " fmt, ap)
+
+void log_buffer_id_data(uint32_t argCount, ...);
+void log_buffer_string_data(const char *format, ...) __attribute__ ((__format__(__printf__, 1, 2)));
+void log_buffer_string_vdata(const char *format, va_list args);
+void log_buffer_string_vdata_critical(const char *format, va_list args, uint8_t lossy);
+void log_assert(const char *format, ...) __attribute__ ((__format__(__printf__, 1, 2)));
+
+// Functions added to test the logging feature
+void log_reset(void);
+uint32_t log_get_bytes_lost(void);
+void log_buffer_data(char *str);
+void log_disable_time_capture(void);
+void log_enable_time_capture(void);
+
+#if defined(NDEBUG) || defined(MBED_ID_BASED_TRACING)
+#define mbed_log_array
```

Comment:
what is this for, empty defines? Also macros are always capital letters. Can this be achieved differently?